src/main/groovy/wooga/gradle/appcenter/tasks/AppCenterUploadTask.groovy

```diff
@@ -182,42 +136,46 @@ class AppCenterUploadTask extends ConventionTask {
         action.execute(buildInfo)
     }
 
-    private Object binary
+    @InputFile
+    final RegularFileProperty binary
 
-    @SkipWhenEmpty
-    @InputFiles
-    protected FileCollection getInputFiles() {
-        if (!binary) {
-            return project.files()
-        }
-        return project.files(binary)
+    void setBinary(String value) {
+        binary.set(project.file(value))
     }
 
-    File getBinary() {
-
-        def files = getInputFiles()
-        if (files.size() > 0) {
-            return files.getSingleFile()
-        }
-        return null
+    void setBinary(File value) {
+        binary.set(value)
     }
 
-    void setBinary(Object value) {
-        binary = value
+    void binary(String value) {
+        setBinary(value)
     }
 
-    AppCenterUploadTask binary(Object binary) {
-        setBinary(binary)
-        this
+    void binary(File value) {
+        setBinary(value)
     }
 
-    @OutputFiles
-    protected FileCollection getOutputFiles() {
-        return project.files(getUploadVersionMetaData())
-    }
+    @Internal
+    protected final DirectoryProperty outputDir
+
+    @OutputFile
+    final Provider<RegularFile> uploadVersionMetaData
+
+    AppCenterUploadTask() {
+        def projectLayout = new ProjectLayout(project)
+        apiToken = project.objects.property(String)
+        owner = project.objects.property(String)
+        buildVersion = project.objects.property(String)
+        releaseId = project.objects.property(Integer)
+        applicationIdentifier = project.objects.property(String)
+        releaseNotes = project.objects.property(String)
+        destinations = project.objects.listProperty(Map)
```

Comment:
This looks funky, just curious, why do you need to do that?

Reply:
The provider API just wraps the properties in a generic wrapper object which can hold either concrete values or other provider objects which all get evaluated at runtime when the value is requested. The properties itself can't no longer be simple java beans (getter/setter) but are `final` objects so they get constructed once for each instance. This code in the constructor is just the bootstrap. In grovvy it is possible to do that directly after the declaration like

```groovy
final Property<String> apiToken = project.objects.property(String)
```

But I personaly like the java way better. This also allows me to chain some provider together.
`project.object` just returns a factory with utility functions to create these generic `Property<T>`. One needs to pass the type of the property since java has not type inference.